[pharrell.mp3] 지하철 서비스 step3 PR 리뷰 부탁드립니다.

pilgyumin opened the pull request:
안녕하세요!
Pharrell(퍼렐_민필규)입니다.

이번 지하철 서비스 프로젝트는 brown이 개발한 코드를 기반으로 진행하였습니다.

실제 현업에서 베이스 코드 위에 새로운 기능을 추가하는 개발 업무가 많이 진행되는 것으로
알고 있어, 해당 부분을 연습하는 느낌으로 진행하였습니다.

이번 프로젝트 진행하면서, 저와 페어가 해결하지 못한 부분이 있어, 질문 드립니다.

subway/auth/infrastructure/AuthenticationPrincipalArgumentResolver.java 34~36 line

        if (!authService.validateToken(token)) {
            return null;
        }

subway/path/application/PathService 59~61 line, 

        public int getLoginMemberAge(LoginMember loginMember) {
            return (loginMember != null) ? loginMember.getAge().intValue() : AGE_DEFAULT;
        }

위의 부분은 경로 검색 기능에서 로그인을 하지 않은 유저가 경로 검색을 요청할 때, 
LoginMember의 정보가 없는 상황을 처리하는 로직입니다.

해당 부분이 이런 방식으로 구현되는 지 여부가 궁금합니다.

위에 질문에 대한 답변 부탁드리겠습니다.
그리고 이번 코드 리뷰도 잘 부탁드리겠습니다.

감사합니다.

pharrell-mp3 committed:
project init

pharrell-mp3 committed:
docs(readme.md) : 기능 구현 목록 작성

pharrell-mp3 committed:
feat(path, line) : 검색 시 최단 거리 경로 구하기

- 역과 구간 전체 데이터를 통해 jgrapht를 사용하여 구현

pharrell-mp3 committed:
feat(path, line) : 출발역과 도착역 사이의 최단 거리 경로 계산

- 검색 시 최단 거리 경로 구하기 (역과 구간 전체 데이터를 통해 jgrapht를 사용하여 구현)
- 검색시 총 거리 구하기
- 여러 노선의 환승도 고려한 검색

pharrell-mp3 committed:
feat(path) : SubwayPathGraph 일급객체 제작

pharrell-mp3 committed:
docs(readme.md) : 기능 구현 목록 작성

pharrell-mp3 committed:
feat(auth) : 토큰 발급 요청 및 응답

- 토큰 발급 요청 및 응답 구현
- 멤버 관리 기능 요청 시 토큰에 대한 로그인멤버 객체 리턴

pharrell-mp3 committed:
feat(auth) : 인증 과정에서 에러 발생 시 401 status로 응답하게 구현

- member 조회 시 Exception이 발생하면 401로 응답

pharrell-mp3 committed:
feat(auth) : 멤버 관리 기능에 토큰 인증 로직 구현

- 토큰에 대한 유효성 검증 로직 구현

pharrell-mp3 committed:
docs(readme.md) : 기능 구현 목록 업데이트

pharrell-mp3 committed:
feat(favorite) : 즐겨찾기 생성 요청 및 응답 구현

- 테이블 스키마 정의 및 요청에 대한 처리 구현

pharrell-mp3 committed:
feat(favorite) : 즐겨찾기 목록 조회 및 삭제 요청 응답 구현

pharrell-mp3 committed:
refactor(favorite) : 코드 리팩토링

- 쿼리문 enum으로 정의
- 코드 포멧팅 및 변수 final 선언

pharrell-mp3 committed:
docs(readme.md) : step3 기능 목록 작성

pharrell-mp3 committed:
feat(path) : 거리별 요금 정책 구현

- 이용 거리 초과 시 추가운임 부과 구현
- 거리별 추가 요금 Enum 클래스 정의

pharrell-mp3 committed:
feat(line) : line 클래스에 extraFare 추가

- 거리별 추가요금 사용을 위한 필드 추가

pharrell-mp3 committed:
feat(line) : 노션별 추가 요금 정책 구현

- Line 클래스에 노선 추가요금 필드 추가
- Line 테이블 스키마에 추가요금 컬럼 추가
- 경로 중 가장 높은 금액의 추가 요금만 적용되는 로직 구현

pharrell-mp3 committed:
refactor(gradle) : 불필요한 의존성 제거

pharrell-mp3 committed:
feat(path) : 연령별 요금 할인 정책

- 연령 Enum 클래스 정의
- 청소년 요금 할인 정책 (운임에서 350원을 공제한 금액의 20% 할인)
- 어린이 요금 할인 정책 (운임에서 350원을 공제한 금액의 50% 할인)

pharrell-mp3 committed:
refactor(auth,common) : 클래스 패키지 변경

- common 패키지에 DataLoader, ExceptionAdvice 클래스 이관
- auth/infrastructure 패키지에 AuthenticationPrincipalConfig, AuthenticationPrincipalArgumentResolver 클래스 이관

pharrell-mp3 committed:
refactor(auth,path) : 클래스의 패키지 변경

pharrell-mp3 committed:
refactor(favorite) : FavoriteResponse 객체 생성에 정적 팩토리 메소드 적용

pharrell-mp3 committed:
refactor(path) : DiscountAge 정의 및 리팩토링

- 할인할 수 있는 나이를 enum 클래스로 정의

pharrell-mp3 committed:
refactor(path) : LoginMember 나이 확인 로직 구현

pharrell-mp3 committed:
refactor(Station) : StationQuery Enum 클래스 정의

pharrell-mp3 committed:
refactor(path) : DistanceExtraFare 내부 리팩토링

- 일반 숫자를 상수화
- 추가요금 계산 시에 사용되는 로직 메소드화

pharrell-mp3 committed:
refactor(Member) : MemberQuery Enum 클래스 정의

pharrell-mp3 committed:
refactor(line) : Section 생성자 리팩토링

pharrell-mp3 committed:
refactor(line) : SectionQuery, LineQuery enum 클래스 정의

pharrell-mp3 committed:
refactor(line) : lineResponse 생성자에 extraFare 적용

pharrell-mp3 committed:
refactor(auth) : TokenResponse 생성자 정적 팩토리 메소드 적용

pharrell-mp3 committed:
refactor(all) : 코드 포멧팅 및 불필요한 import 제거

adeldel commented on README.md:
```diff
@@ -38,6 +38,49 @@ npm run dev
+* [x] 연령 Enum 클래스 정의
+* [x] 청소년 요금 할인 정책 (운임에서 350원을 공제한 금액의 20% 할인)
+* [x] 어린이 요금 할인 정책 (운임에서 350원을 공제한 금액의 50% 할인)
+
```
상세한 구현 요구사항 정리해주셔서 감사합니다. 💯

pilgyumin replied:
넵 백점 감사합니다!!

adeldel commented on src/main/java/subway/auth/application/AuthService.java:
```diff
@@ -1,4 +1,48 @@
+
+    public TokenResponse createToken(TokenRequest tokenRequest) {
+        if (!checkValidLogin(tokenRequest.getEmail(), tokenRequest.getPassword())) {
+            throw new InvalidLoginException();
```
에러 메시지도 넘겨주시면 연동시 어떤 에러인지 명확하게 알수 있어요. 해당 에러가 어느 시점에(또는 어느 상황에) throw했는지 알면 문제 해결에 도움이 됩니다.

pilgyumin replied:
네 에러 메세지를 설정하는 방법으로 수정하겠습니다!

adeldel commented on src/main/java/subway/auth/exception/InvalidLoginException.java:
```diff
@@ -0,0 +1,4 @@
+package subway.auth.exception;
+
+public class InvalidLoginException extends RuntimeException{
```
에러 메시지를 남기려면 여기도 생성자를 추가해주셔야겠지요?

pilgyumin replied:
넵 생성자 추가하여 반영하겠습니다!

adeldel commented on src/main/java/subway/auth/infrastructure/AuthenticationPrincipalArgumentResolver.java:
```diff
@@ -20,10 +24,17 @@ public boolean supportsParameter(MethodParameter parameter) {
+
+        String token = AuthorizationExtractor.extract((HttpServletRequest) webRequest.getNativeRequest());
+        if (!authService.validateToken(token)) {
+            return null;
```
여기서 `LoginMember`를 null을 리턴하면  파라미터로 LoginMember를 받아서 사용하는 컨트롤러에서 어떤 일이 일어날까요? 널체크를 하지 않아서 `loginMember.getId() `같은 로직에서 NPE가 발생할 수 있습니다.
인증이 필요한 곳은 모두 loginMember를 파라미터로 쓰고 있기 때문에  여기서 널을 리턴할게 아니라 디폴트값으로라도 `LoginMember`를 생성해서 넘겨주는게 좋을 것 같아요. 또는 컨트롤러까지 넘어갈 필요가 없다면 에러처리를 해도 좋구요.
resolveArgument 의 책임이 정상 회원정보를 돌려주는 것만으로도 loginMember를 파라미터로 사용하는 곳에서는 안전하게 쓸수 있어요.

adeldel replied:
`최단경로조회` 기능에서는 필수가 아니라면 널보다는 디폴트값이라도 채워서 `LoginMember` 객체를 리턴해주는 방식이 좋을 것 같아요.

pilgyumin replied:
네 LoginMember 클래스 내에 defaultLoginMember 정보를 갖고 있게 하겠습니다.
그래서 토큰의 유효성 여부에 따라서 토큰에 대한 LoginMember 와 디폴트값으로 리턴할 수 있도록 구현하겠습니다.

adeldel commented on src/main/java/subway/auth/infrastructure/LoginInterceptor.java:
```diff
@@ -0,0 +1,29 @@
+    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) {
+        String accessToken = AuthorizationExtractor.extract(request);
+        if (accessToken == null || !authService.validateToken(accessToken)) {
+            throw new InvalidLoginException();
```
`HandlerInterceptor` 를 구현해주셨어요 👍  
회원관리와 즐겨찾기할때에는 유효한 회원만 사용할 수 있도록 처리해 주셨어요. 시도는 좋습니다. 
그런데, `AuthenticationPrincipalArgumentResolver` 의 `resolveArgument` 로직과 겹치는 느낌입니다. 
여기서 `validateToken`을 통과하지 못한다면 `resolveArgument` 에서 `InvalidLoginException` 을 던져도 괜찮지 않을까요?
어차피 컨트롤러까지 요청이 진입하지 못하고 에러처리를 해야 한다면요.

adeldel replied:
아,  리뷰를 하면서 쭈욱 따라가다보니  `최단거리 경로 조회`시에는 로그인멤버가 필수가 아니었네요.
LoginInterceptor 가 필수인 컨트롤러에 진입전에 validate 체크하기 위한 기능을 하고 있어서 이대로 두어도 좋습니다. 👍

pilgyumin replied:
넵 감사합니다!!

adeldel commented on src/main/java/subway/common/ExceptionAdvice.java:
```diff
@@ -0,0 +1,15 @@
+    @ExceptionHandler(InvalidLoginException.class)
+    public ResponseEntity<String> catchInvalidLoginException() {
+        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();
+    }
```
`ControllerAdvice` 를 사용해주셨어요 👍 
헌데 다른 컨트롤러는 `ExceptionHandler` 를 컨트롤러 클래스에서 관리하고 있어요.
다른 클래스의 `ExceptionHandler` 들을 여기로 모으는건 어떨까요?

pilgyumin replied:
오 넵넵 다른 컨트롤러의 ExceptionHandler가 존재하는 지 몰랐었습니다! ExceptionAdvice 클래스로 이관하겠습니다!

adeldel commented on src/main/java/subway/common/ExceptionAdvice.java:
```diff
@@ -0,0 +1,15 @@
+package subway.common;
```
`common -> exception` 패키지는 어떨까요?

pilgyumin replied:
넵 common 패키지로 이관하겠습니다.

adeldel commented on src/main/java/subway/favorite/application/FavoriteService.java:
```diff
@@ -0,0 +1,40 @@
+package subway.favorite.application;
```
`application -> service` 는 어떨까요? 여타 다른 기능들도요.

pilgyumin replied:
넵 알겠습니다. 패키지명을 좀 더 직관적으로 변경하겠습니다!

adeldel commented on src/main/java/subway/auth/application/AuthService.java:
```diff
@@ -1,4 +1,48 @@
+    }
+
+    public TokenResponse createToken(TokenRequest tokenRequest) {
+        if (!checkValidLogin(tokenRequest.getEmail(), tokenRequest.getPassword())) {
```
요청으로 들어온 tokenRequest 에 값이 있는지 여부도 체크를 하면 디비 조회전에 Invalid한지 알수 있어요. (디비 쿼리수 줄이기)

pilgyumin replied:
넵 tokenRequest 내에 isEmpty()를 구현하겠습니다!

adeldel commented on src/main/java/subway/favorite/dao/FavoriteQuery.java:
```diff
@@ -0,0 +1,22 @@
+package subway.favorite.dao;
+
+public enum FavoriteQuery {
```
enum 클래스로 쿼리를 모아주셨네요. 👍 
저는 이렇게 써본적은 없지만  😅  어차피 enum이 상수모음이기 때문에 괜찮은 것 같아요.

pilgyumin replied:
오 넵 칭찬 감사합니다!!😄

adeldel commented on src/main/java/subway/favorite/ui/FavoriteController.java:
```diff
@@ -1,8 +1,51 @@
+
+    @DeleteMapping("/{id}")
+    public ResponseEntity deleteFavorite(@PathVariable Long id) {
+        favoriteService.deleteFavorite(id);
```
즐찾을 삭제할때에도 유효한 회원인지 판단해야 하지 않을까요? 아이디값만 알면 누구나 남의 즐찾도 삭제가능한 상태가 됩니다.

pilgyumin replied:
네 타인의 즐겨찾기 삭제 요청 시 예외가 발생하게 구현하겠습니다.

adeldel commented on src/main/java/subway/path/ui/PathController.java:
```diff
@@ -1,5 +1,35 @@
+
+        return ResponseEntity
+                .ok()
+                .body(pathService.getShortestPathResponse(source, target, pathService.getLoginMemberAge(loginMember)));
```
질문주신 내용이 이 로직인데요. 애초에 `loginMember`를 널을 리턴하지 않으면 널체크를 여기서 하지 않아도 될것 같아요.
그리고 로그인 유저가 굳이 경로 조회시 필수는 아니라면  
`AuthenticationPrincipal` 어노테이션에 `required = false`로 하는건 어떨까요?

pilgyumin replied:
네 해당 부분 수정하겠습니다!

adeldel commented on src/main/java/subway/path/domain/DistanceExtraFare.java:
```diff
@@ -0,0 +1,62 @@
+package subway.path.domain;
+
+public enum DistanceExtraFare {
```
거리에 따른 요금계산을 enum 으로 구현해서 아주 깔끔합니다. 👍 👍 👍

pilgyumin replied:
오 넵 감사합니다!!😀

adeldel commented on src/main/java/subway/path/domain/DiscountAge.java:
```diff
@@ -0,0 +1,45 @@
+package subway.path.domain;
+
+public enum DiscountAge {
```
나이에 따른 요금계산 👍 👍 👍

pilgyumin replied:
칭찬 감사합니다!!😃

adeldel reviewed (changes requested):
안녕하세요. pharrell 리뷰가 다소 늦어진 점 죄송해요 🙏 
소스가 커밋단위로 잘 쪼개주셔서 기능구현 스탭을 잘 볼수 있었어요. 💯 
그리고 기능 요구사항도 잘 정리해주셔서 리뷰에 도움이 되었답니다.
전체적 구조 및 도메인객체 분리, 쿼리분리 등 전반적인 로직을 깔끔하기 잘 작성해주셨어요. 
특히 나이나 거리에 따른 요금계산 로직을 enum으로 분리하고 책임을 잘 분리해주셔서 인상적이었습니다.
쪼오금 아쉬운 점은 유닛테스트 작성이 부족하고, validate 체크가 필요한 부분이 보여 댓글로 남겨드렸어요.
수정 후 재요청 부탁드립니다. 구현하느라 수고 많으셨어요 🎉

pharrell-mp3 committed:
refactor(auth) : 커스텀 Exception 내 에러메세지 정의 및 Exception 분리

- Exception 내 에러메세지 설정할 수 있도록 생성자 구현
- InvalidLoginException과 InvalidTokenException 분리

pharrell-mp3 committed:
refactor(common) : exception 패키지 이관

- auth 패키지 내에 있던 exception 패키지를 common으로 이관

pharrell-mp3 committed:
refactor(path, auth, member) : 토큰의 유효성에 대한 LoginMember 객체 생성 처리 구현

- 유효한 토큰일 때, 토큰에 대한 LoginMember 반환
- 유효하지 않은 토큰일 때, defaultLoginMember 반환

pharrell-mp3 committed:
refactor(common) : 401 에러로 응답을 주는 exception 추가

pharrell-mp3 committed:
refactor(common, line, station) : exceptionHandler 메소드를 ExceptionAdvice 클래스로 이관

pharrell-mp3 committed:
refactor(all) : 패키지명 변화

- application -> service
- ui -> controller

pharrell-mp3 committed:
refactor(auth) : TokenRequest 내에 isEmpty 구현

- email, password 둘 중 하나가 비어있을 경우 true 리턴

pharrell-mp3 committed:
refactor(favorite) : 타인의 즐겨찾기 삭제 요청에 대한 예외 처리

- 유저가 저장한 즐겨찾기일 때만 삭제 가능하게 로직 구현

pharrell-mp3 committed:
feat(path) : 나이 별 요금 할인 테스트

pharrell-mp3 committed:
feat(path) : 거리별 추가 요금 테스트

adeldel commented on src/main/java/subway/member/domain/LoginMember.java:
```diff
@@ -1,10 +1,23 @@
+
+    static {
+        defaultLoginMember = new LoginMember(ID_DEFAULT, EMAIL_DEFAULT, AGE_DEFAULT);
+    }
```
static 구문을 따로 빼실 필요는 없어요. 멤버변수에서 바로 초기화 하시면 됩니다. 
```suggestion
private static LoginMember defaultLoginMember = new LoginMember(ID_DEFAULT, EMAIL_DEFAULT, AGE_DEFAULT);

pilgyumin replied:
넵 알겠습니다!

adeldel commented on src/main/java/subway/member/domain/LoginMember.java:
```diff
@@ -1,10 +1,23 @@
+        defaultLoginMember = new LoginMember(ID_DEFAULT, EMAIL_DEFAULT, AGE_DEFAULT);
+    }
+
     public LoginMember() {
```
사용하지 않는다면 불필요한 생성자는 제거해주세요.

pilgyumin replied:
넵 제거하겠습니다!

adeldel commented on src/main/java/subway/favorite/service/FavoriteService.java:
```diff
@@ -30,7 +31,13 @@ public FavoriteResponse createFavorite(FavoriteRequest favoriteRequest, Long mem
+                .stream()
+                .filter(favorite -> favorite.getId().equals(id))
+                .findFirst()
+                .orElseThrow(() -> new IllegalStateException("타인의 즐겨찾기는 삭제할 수 없습니다."));
```
👍 👍

pilgyumin replied:
칭찬 감사합니다!!

adeldel commented on src/main/java/subway/auth/infrastructure/AuthenticationPrincipalArgumentResolver.java:
```diff
@@ -20,10 +24,19 @@ public boolean supportsParameter(MethodParameter parameter) {
+            member = authService.findMemberByToken(token);
+        }
+
+        return LoginMember.of(member);
```
validateToken이 아니라면 member가 null이  들어갈 수 있어요.
변수는 최소한의  scope에서만 사용하는게 어떨까요? 디비에서 조회해서 early return 하고, validateToken 이 아니라면 default LoginMember 를 돌려주도록요. (defaultLoginMember 접근자도 변경해야겠네요.)
``` suggestion
       if (authService.validateToken(token)) {
            Member member = authService.findMemberByToken(token);
            return LoginMember.of(member);
        }

        return LoginMember.defaultLoginMember;

pilgyumin replied:
아 넵 알겠습니다! 해당 부분 반영하겠습니다!

adeldel commented on src/main/java/subway/auth/service/AuthService.java:
```diff
@@ -0,0 +1,56 @@
+            return false;
+        }
+        return true;
+    }
```
여기서 리턴타입은 필요없어 보입니다. StringBuilder 를 넘겨줄 필요도 없고요. 
다양한 exception 이나 에러메시지를 처리하고 싶다면, 상황별로 에러를 throw 해주면 될것 같아요.
```suggestion
    private void checkValidLogin(String email, String password) {
        try {
            if (!memberDao.findByEmail(email)
                    .getPassword()
                    .equals(password)) {
                throw new InvalidLoginException(InvalidLoginException.EMAIL_PASSWORD_MISMATCH);
            }
        } catch (EmptyResultDataAccessException erdae) {
            throw new InvalidLoginException(InvalidLoginException.EMAIL_NOT_EXIST);
        }
    }

pilgyumin replied:
네 void로 반환형을 변경하고 분기처리에 맞는 예외를 throw하도록 구현하겠습니다.

adeldel commented on src/main/java/subway/auth/service/AuthService.java:
```diff
@@ -0,0 +1,56 @@
+        }
+        String accessToken = jwtTokenProvider.createToken(tokenRequest.getEmail());
+        return TokenResponse.of(accessToken);
+    }
```
checkValidLogin에서 리턴타입을 없앤다면 여기도 이렇게 변경해볼 수는 있어요. (참고만 해주세요.) 
```suggestion
    public TokenResponse createToken(TokenRequest tokenRequest) {
        if (tokenRequest.isEmpty()) {
            throw new IllegalArgumentException("tokenRequest invalid");
        }
        checkValidLogin(tokenRequest.getEmail(), tokenRequest.getPassword());
        String accessToken = jwtTokenProvider.createToken(tokenRequest.getEmail());
        return TokenResponse.of(accessToken);
    }

pilgyumin replied:
넵 알겠습니다!

adeldel reviewed (approved):
안녕하세요 패럴 😄  리뷰반영을 아주 잘 해주셨어요. 
다만 쪼금 아쉬운 부분이 있어서 제가 의견을 드린 코멘트가 있는데 (사람마다 스타일이 다르고, 제가 정답은 아니므로) 참고해주시면 됩니다. 
이번 스탭은 머지해놓을께요. 정말 수고많으셨습니다. 🎉
